lib/puppetserver/ca/generate_action.rb

```diff
@@ -0,0 +1,123 @@
+
+Options:
+BANNER
+
+      def initialize(logger)
+        @logger = logger
+      end
+
+      def run(input)
+        create_csr
+        puts "\nTake the CSR that was just generated (also written to tmp/req.pem) and submit it\nfor signing by your public key infrastructure (PKI). The signing procedure\ndepends on the PKI that your organization deploys; ask your PKI administrators\nfor more information."
+        return 0
+      end
+
+
+      # Generates an intermediate CA certificate request.
+      # Writes CSR to tmp/req.pem and prints the same CSR to stdout.
+      def create_csr
+        GenCACSR.new.run
+      end
+
+      def parse(cli_args)
+        parser, inputs, unparsed = parse_inputs(cli_args)
+
+        if !unparsed.empty?
+          @logger.err 'Error:'
+          @logger.err 'Unknown arguments or flags:'
+          unparsed.each do |arg|
+            @logger.err "    #{arg}"
+          end
+
+          @logger.err ''
+          @logger.err parser.help
+
+          exit_code = 1
+        else
+          exit_code = nil
+        end
+
+        return inputs, exit_code
+      end
+
+      def parse_inputs(inputs)
+        parsed = {}
+        unparsed = []
+
+        parser = self.class.parser(parsed)
+
+        begin
+          parser.order!(inputs) do |nonopt|
+            unparsed << nonopt
+          end
+        rescue OptionParser::ParseError => e
+          unparsed += e.args
+          unparsed << inputs.shift unless inputs.first =~ /^-{1,2}/
+          retry
+        end
+
+        return parser, parsed, unparsed
+      end
+
+      def self.parser(parsed = {})
+        OptionParser.new do |opts|
+          opts.banner = BANNER
+          opts.on('--help', 'Display this generate specific help output') do |help|
+            parsed['help'] = true
+          end
+          opts.on('--version', 'Output the version') do |v|
+            parsed['version'] = true
+          end
+        end
+      end
+    end
+
+    class GenCACSR < Puppet::Application
+      run_mode :master
+
+      def main
+        root_cert = OpenSSL::X509::Certificate.new(File.read(Puppet[:cacert]))
+        ca_key = OpenSSL::PKey::RSA.new(File.read(Puppet[:cakey]))
+
+        csr = OpenSSL::X509::Request.new
+        csr.version = 0
+        csr.subject = root_cert.subject
+        csr.public_key = root_cert.public_key
+
+        ef = OpenSSL::X509::ExtensionFactory.new
+
+        ext_req = OpenSSL::X509::Attribute.new("extReq",
+          OpenSSL::ASN1::Set([OpenSSL::ASN1::Sequence([
+            ef.create_extension("keyUsage", "cRLSign, keyCertSign", true),
+            ef.create_extension("basicConstraints", "CA:TRUE", true)])]))
+
+        csr.add_attribute(ext_req)
+        csr.sign(ca_key, OpenSSL::Digest::SHA256.new)
+
+        directory_name = "tmp"
+        Dir.mkdir(directory_name) unless File.exists?(directory_name)
+
+        File.open("#{directory_name}/req.pem", "w") do |f|
+          f.write(csr.to_s)
```

Comment:
Do we want to save their csr here?  or where would be good?  It also prints to stdout.